In [ ]:
import pandas as pd
import os

In [ ]:
os.chdir('..')
df = pd.read_csv('scripts/hyperparameters_n_results.csv')
df.head()

,run_id,run_name,min_samples_leaf,criterion,optimized_metric,n_estimators,max_features,max_depth,bootstrap,min_samples_split,max_samples,cohen_kappa,f1_score,matthews_corrcoef,precision,recall,roc_auc_score
0,77853896cb4d4883a1b698915d7c741f,RFC_with_sqrt_n_False_bootstrap_n_roc_auc_score,44,entropy,roc_auc_score,215,sqrt,10,False,457,NaN,0.666261,0.783684,0.677024,0.833507,0.776728,0.956598
1,0fd2d22a01204ac5963dd5b71af5c8d4,RFC_with_sqrt_n_False_bootstrap_n_matthews_cor...,231,gini,matthews_corrcoef,208,sqrt,10,False,539,NaN,0.663720,0.782924,0.673558,0.829505,0.775044,0.954281
2,bfb2cdeb64df4666b8724aa8fcd974b6,RFC_with_sqrt_n_False_bootstrap_n_cohen_kappa,214,gini,cohen_kappa,50,sqrt,10,False,948,NaN,0.657128,0.778102,0.667400,0.826513,0.770390,0.952016
3,5631a7dbc5a343aca68d07031b149d5f,RFC_with_sqrt_n_False_bootstrap_n_precision,231,entropy,precision,209,sqrt,10,False,235,NaN,0.660279,0.779375,0.671381,0.830726,0.772561,0.955608
4,0cc5c6d902454b40b82e9decad0ae8da,RFC_with_sqrt_n_False_bootstrap_n_recall,138,entropy,recall,254,sqrt,10,False,1092,NaN,0.649998,0.771681,0.661701,0.824962,0.765476,0.952659


In [10]:
df_ranks = pd.DataFrame({
    'run_name': df['run_name'],
    'f1_rank': df['f1_score'].rank(ascending=False).astype(int),
    'precision_rank': df['precision'].rank(ascending=False).astype(int),
    'recall_rank': df['recall'].rank(ascending=False).astype(int),
    'kappa_rank': df['cohen_kappa'].rank(ascending=False).astype(int),
    'mcc_rank': df['matthews_corrcoef'].rank(ascending=False).astype(int),
    'roc_auc_rank': df['roc_auc_score'].rank(ascending=False).astype(int)
})

df_ranks.head(12)

,run_name,f1_rank,precision_rank,recall_rank,kappa_rank,mcc_rank,roc_auc_rank
0,RFC_with_sqrt_n_False_bootstrap_n_roc_auc_score,2,1,2,1,1,1
1,RFC_with_sqrt_n_False_bootstrap_n_matthews_cor...,3,4,3,3,3,3
2,RFC_with_sqrt_n_False_bootstrap_n_cohen_kappa,6,6,6,6,6,9
3,RFC_with_sqrt_n_False_bootstrap_n_precision,5,2,5,5,4,2
4,RFC_with_sqrt_n_False_bootstrap_n_recall,10,7,10,10,9,6
5,RFC_with_sqrt_n_False_bootstrap_n_f1,4,5,4,4,5,5
6,RFC_with_sqrt_n_True_bootstrap_n_roc_auc_score,9,8,9,8,8,7
7,RFC_with_sqrt_n_True_bootstrap_n_matthews_corr...,11,11,11,11,11,8
8,RFC_with_sqrt_n_True_bootstrap_n_cohen_kappa,1,3,1,2,2,4
9,RFC_with_sqrt_n_True_bootstrap_n_precision,8,9,8,9,10,11


Тут мы видим, что самые лучшие модели это:
1. RFC_with_sqrt_n_False_bootstrap_n_roc_auc_score (естественно)
2. RFC_with_sqrt_n_True_bootstrap_n_cohen_kappa
3. RFC_with_sqrt_n_False_bootstrap_n_matthews_corrcoef
4. RFC_with_sqrt_n_False_bootstrap_n_precision

In [11]:
best_models = [
    'RFC_with_sqrt_n_False_bootstrap_n_roc_auc_score',
    'RFC_with_sqrt_n_True_bootstrap_n_cohen_kappa',
    'RFC_with_sqrt_n_False_bootstrap_n_matthews_corrcoef',
    'RFC_with_sqrt_n_False_bootstrap_n_precision'
]

df[df['run_name'].isin(best_models)].head()

,run_id,run_name,min_samples_leaf,criterion,optimized_metric,n_estimators,max_features,max_depth,bootstrap,min_samples_split,max_samples,cohen_kappa,f1_score,matthews_corrcoef,precision,recall,roc_auc_score
0,77853896cb4d4883a1b698915d7c741f,RFC_with_sqrt_n_False_bootstrap_n_roc_auc_score,44,entropy,roc_auc_score,215,sqrt,10,False,457,NaN,0.666261,0.783684,0.677024,0.833507,0.776728,0.956598
1,0fd2d22a01204ac5963dd5b71af5c8d4,RFC_with_sqrt_n_False_bootstrap_n_matthews_cor...,231,gini,matthews_corrcoef,208,sqrt,10,False,539,NaN,0.663720,0.782924,0.673558,0.829505,0.775044,0.954281
3,5631a7dbc5a343aca68d07031b149d5f,RFC_with_sqrt_n_False_bootstrap_n_precision,231,entropy,precision,209,sqrt,10,False,235,NaN,0.660279,0.779375,0.671381,0.830726,0.772561,0.955608
8,8a6fe937e1ea4b20ad7dbfbc78bcd3a2,RFC_with_sqrt_n_True_bootstrap_n_cohen_kappa,27,gini,cohen_kappa,200,sqrt,10,True,418,0.948009,0.666043,0.784618,0.675524,0.829512,0.777023,0.954133


Тут представлены гиперпараметры 4-х топовых моделей. Можно заметить следующее:
1. В двух моделей min_samples_leaf абсолютно идентичны (231), в то же время у других двух моделях значения схожи (44 и 27), но категорически отличаются от двух предыдущих
2. По критерию разбиения тоже есть различия: две модели используют энтропию, другие две - gini
3. Метрика, которая каждая модель улучшает, тоже разная, причем у всех
4. n_estimators у всех приблизительно одинаковый
5. У всех одинаковая глубина, она равна 10
6. Только одна модель использует bootstrap
7. min_samples_split лежит в диапазоне от 235 до 539

Из всего этого можно выбрать улучшаемую метрику roc_auc_score и следующие диапазоны гиперпараметров:
1. min_samples_leaf - значения меньше 100, в районе 40
2. criterion='entropy'
3. n_estimators - [190, 240]
4. max_features='sqrt'
5. max_depth - [9, 11]
6. bootstrap=False
7. min_samples_split - [417, 497]